sparse/client.go

```diff
@@ -138,14 +180,80 @@ func (client *syncClient) syncContent() error {
 	return err
 }
 
+func (client *syncClient) isLocalAndRemoteDiskFilesIdentical() bool {
+	if client.recordedChangeTime == "" || client.recordedChecksumMethod == "" || client.recordedChecksum == "" {
+		return false
+	}
+
+	remain, err := client.isLocalDiskFileRemain()
+	if err != nil {
+		log.WithError(err).Warnf("Failed to check if local file %v is changed", client.sourceName)
+		return false
+	}
+	if !remain {
+		return false
+	}
+
+	remoteRecordedChecksum, err := client.getRemoteDiskFileRecordedChecksum()
+	if err != nil {
+		log.WithError(err).Warnf("Failed to get checksum for remote file %v", client.sourceName)
+		return false
+	}
+
+	return client.recordedChecksum == remoteRecordedChecksum
+}
+
+func (client *syncClient) isLocalDiskFileRemain() (bool, error) {
+	currentChangeTime, err := util.GetFileChangeTime(client.sourceName)
+	if err != nil {
+		return false, err
+	}
+	return currentChangeTime == client.recordedChangeTime, nil
+}
+
+func (client *syncClient) getRemoteDiskFileRecordedChecksum() (string, error) {
+	metadata, err := client.getServerRecordedMetadata()
+	if err != nil {
+		return "", err
+	}
+
+	var info types.SnapshotHashInfo
+	if err := json.Unmarshal(metadata, &info); err != nil {
+		return "", errors.Wrap(err, "failed to unmarshal hash info")
+	}
+
+	return info.Checksum, nil
+}
+
+func getLocalDiskFileChangeTimeAndChecksum(sourceName string) (string, string, string, error) {
```

Comment:
nit: when having more than 2 return vars, I would suggest having named return vars to make the protocol clear.